### PR TITLE
SQL 2008 support, Replication feature detection, xSQLNetwork resource (closes #6 and #16, partialy solves #14, fixes #7(2 and 3))

### LINKED ISSUE #6: SourcePath and UpdateEnabled
Hi Team
Having an issue with this module, not sure if it's my issue or issue with the code.
Using the sample included I cannot get the sourcepath or updateenabled/updatesource parameters to work.
Digging through the various modules I found that there were hard coded values in xSQLServerSetup.psm1 and they were not getting replaced with my parameter values.
I removed the hard coded values from the psm1 file and it correctly picked up my parameter values.
Is this expected behaviour?
Cheers

### LINKED ISSUE #7: Issues during using this DSCResource
Hi have use the xSQLServer Resource to install SQLExpress Adv (DEU) on Server Core here are my Experience, maybe that helps the developer :

1. If i started with the Resource on a german language Server, there is a Issue with LocalizedData in the xPDT.psm1. I copy en-Us folder to a de-DE and the Issues are gone. So it would be better to check this, so the others in world could use the resource
2. My Installation-Resource (setup.exe) is on the UNC-Path in the network, the Resource script called the function GetSQLVersion, but there is no net_use command first to make a secure Connection to the UNC-Source. So the function call failed and the resource execution failed.
3. If you use netuse from the xPDT.psm1 and the Setup Credential has a complex Password (maybe double Quote inside), the netuse command failed and the connection couldn’t be established.  Maybe it is better to use with pinvoke the WNetAddConnection2 or NetUseAdd methods. I use the implementation from http://stackoverflow.com/questions/1477328/calling-wnetaddconnection2-from-powershell and it worked fine.
4. The second problem with complex password is in the execution of the setup.exe. If you defined a complex SQLService-Account-Password, then the Installation failed (maybe double Quote inside) with wrong password. So it would be better to escaped the arguments. A good starting point for this is http://blogs.msdn.com/b/twistylittlepassagesallalike/archive/2011/04/23/everyone-quotes-arguments-the-wrong-way.aspx. If I use the rules in the blog the complex password is correct committed to the setup.exe
5. If you check the features, then the check is CASEsensitiv, so if you not take the correct CASE in the Resource-Property then the installed Feature is not recognized correct. So maybe it is better to make a IGNORE-CASE compare. But this issue is addressed on https://github.com/PowerShell/xSQLServer/commits/dev ‘Resolved problem with $Features due to Contains method being case sen…’.

I hope this information helps, maybe you already know this all ;-), but I spend a lot of time and I thought it would be a good idea to share this information
best regards jens m.

### PR DESCRIPTION
**Pull request by Arturas-K**
Hi,

I'm using DSC and xSqlServer module to automate installation of my environments.
While using this module I have encountered several issues and missing functionality.
So I've made several changes to existing modules and added new one to solve what I needed to accomplish.

Wasn't sure on new module class name should I be using MSFT_ prefix or not, current implementation is without a prefix, let me know if you want me to add it.

Bellow is a summary of changes, this information is also added to readme.md Unreleased section as per contribution guidance:
* xSQLServerSetup:
  - Added support for SQL Server 2008 R2 installation
  - Removed default values for parameters, to avoid compatibility issues and setup errors
  - Added Replication sub feature detection
  - Added setup parameter BrowserSvcStartupType
* xSQLServerFirewall
  - Removed default values for parameters, to avoid compatibility issues
  - Updated firewall rule name to not use 2012 version, since package supports 2008, 2012 and 2014 versions
* xSQLServerNetwork
  - Added new resource that configures network settings.
  - Currently supports only tcp network protocol
  - Allows to enable and disable network protocol for specified instance service
  - Allows to set custom or dynamic port values

* Updated example files to use correct DebugMode parameter value ForceModuleImport, this is not boolean in WMF 5.0 RTM

### CODE REVIEW

**Comment by NarineM** [comment]
Can you merge this with #19 that already address some of your concerns and send a ne review?

**Comment by Arturas-K** [comment]
Hi, i have merged my branch with dev following #19 PR.

Almost all of my changes are still relevant.

I have one question @aultt : since #19 introduced $SourceCredentials to xSQLServerSetup and in Set function it copies source to temp folder, i don't see any use of the old $SetupCredentials anymore, it looks to me that we introduced new parameter and made obsolete old one.

Only place $SetupCredentials is referenced is in sysadmin part, which can be achieved using $SQLSysAdminAccounts parameter;

Am i missing something else?

**Comment by NarineM** [comment]
@Arturas-K , thank you for merging the changes. Give me a day to review these changes.
As for your $SetupCred vs $SourceCred question.. These are not duplicates. $SourceCred purpose is to provide credentials for accessing the repository/share/store where the sql installer resides (example net use creds). $SetupCreds are the credentials for installing the sql server on a machine.

**Comment by NarineM** [comment]
Hi @Arturas-K,
Thanks for your contribution!

 In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.

File: DSCResources/MSFT_xSQLServerFirewall/MSFT_xSQLServerFirewall.psm1 [hunk 1]
```diff
@@ -14,7 +14,7 @@ function Get-TargetResource
         $SourcePath = "$PSScriptRoot\..\..\",
 
         [System.String]
-        $SourceFolder = "Source",
+        $SourceFolder = "Source",
```
Comment (NarineM) [inline DSCResources/MSFT_xSQLServerFirewall/MSFT_xSQLServerFirewall.psm1]: No changes in this line, please remove it from the commit.

File: DSCResources/MSFT_xSQLServerFirewall/MSFT_xSQLServerFirewall.psm1 [hunk 2]
```diff
@@ -174,7 +174,7 @@ function Set-TargetResource
...
```
Comment (NarineM) [inline DSCResources/MSFT_xSQLServerFirewall/MSFT_xSQLServerFirewall.psm1]: no change in this line. please remove from the commit.

File: DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1 [hunk 3]
```diff
@@ -43,10 +43,10 @@ function Get-TargetResource
         $PID,
 
         [System.String]
-        $UpdateEnabled = "True",
```
Comment (NarineM) [inline DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1:46]: Why this default is removed?

File: DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1 [hunk 4]
```diff
@@ -176,6 +180,18 @@ function Get-TargetResource
         $SQLSvcAccountUsername = (Get-WmiObject -Class Win32_Service | Where-Object {$_.Name -eq $DBServiceName}).StartName
         $AgtSvcAccountUsername = (Get-WmiObject -Class Win32_Service | Where-Object {$_.Name -eq $AgtServiceName}).StartName
         $FullInstanceID = (Get-ItemProperty -Path 'HKLM:\SOFTWARE\Microsoft\Microsoft SQL Server\Instance Names\SQL' -Name $InstanceName).$InstanceName
+
+        #test if Replication sub component is configured for this instance
+        Write-Verbose "Detecting replication feature"
+        Write-Verbose "Reading registry key HKLM:\SOFTWARE\Microsoft\Microsoft SQL Server\$FullInstanceID\ConfigurationState"
+        $IsReplicationInstalled = (Get-ItemProperty -Path "HKLM:\SOFTWARE\Microsoft\Microsoft SQL Server\$FullInstanceID\ConfigurationState").SQL_Replication_Core_Inst
```
Comment (NarineM) [inline DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1]: Please update local variable names to start with a lowercase. See the guidelines: https://github.com/PowerShell/DscResources/blob/TravisEz13-patch-1/StyleGuidelines.md#variable-names-should-use-camelcase

File: DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1 [hunk 5]
```diff
@@ -563,6 +611,12 @@ function Set-TargetResource
         "InstallSharedWOWDir",
         "InstanceDir"
     )
+
+    if($BrowserSvcStartupType -ne $null)
```
Comment (NarineM) [inline DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1:615]: local variable names should start with a lowercase (as per https://github.com/PowerShell/DscResources/blob/TravisEz13-patch-1/StyleGuidelines.md#variable-names-should-use-camelcase)

File: DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1 [hunk 6]
```diff
@@ -758,7 +812,7 @@ function Test-TargetResource
         $SourcePath = "$PSScriptRoot\..\..\",
 
         [System.String]
-        $SourceFolder = "Source",
+        $SourceFolder = "Source",
```
Comment (NarineM) [inline DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1]: No change in this line. please remove from the commit.

File: DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1 [hunk 7]
```diff
@@ -874,7 +928,11 @@ function Test-TargetResource
         $ASConfigDir,
 
         [System.Management.Automation.PSCredential]
-        $ISSvcAccount
+        $ISSvcAccount,
+
+        [System.String]
+        [ValidateSet("Automatic", "Disabled", "Manual")]
+        $BrowserSvcStartupType
```
Comment (NarineM) [inline DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1:935]: What is the default for BrowserSvcStartupType param?

File: DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1 (hunk 3 above)
Comment (Arturas-K) [inline DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1:46]: I have described the issue in #16 with regarding current defaults.
In addition 2008 R2 setup doesn't support these, so by just removing defaults they no longer passed in as command parameters, this way you can use this resource for 2008 installation.

Other option would be to check version and remove $UpdateEnabled and $UpdateSource parameter from array before running setup.

SQL Server setup has default value of $true anyway, so by removing this in Resource effectively doesn't change the behaviour as setup default takes affect.

https://msdn.microsoft.com/en-us/library/ms144259(v=sql.110).aspx

File: DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1 (hunk 7 above)
Comment (Arturas-K) [inline DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1:935]: By default Browser service is in disabled mode after installation.
If you are installing named instance and want to use dynamic ports Browser service needs to be running.

I don't think we need any default value in Resource level, if this is omitted then it is not passed in to setup.exe so behaviour stays same. So I think this change is backwards compatible and is safe to add.

**Comment by Arturas-K** [comment]
Hi, @NarineM 

Thanks for your review.
I have signed CLA, fixed line endings and update local variables to use camelCase style.

Let me know if you need me to fix anything else.

File: DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1 (hunk 7 above)
Comment (NarineM) [inline DSCResources/MSFT_xSQLServerSetup/MSFT_xSQLServerSetup.psm1:935]: Sounds good.

File: README.md [hunk 8]
```diff
@@ -234,8 +242,21 @@ Please check out common DSC Resources [contributing guidelines](https://github.c
...
->>>>>>> cd039bfd72f5a433a527cef05634382050a4e356
+  - Corrected bug in GetFirstItemPropertyValue to correctly handle registry keys with only one value.
+  - Added support for SQL Server 2008 R2 installation
+  - Removed default values for parameters, to avoid compatibility issues and setup errors
+  - Added Replication sub feature detection
+  - Added setup parameter BrowserSvcStartupType
+* xSQLServerFirewall
+  - Removed default values for parameters, to avoid compatibility issues
+  - Updated firewall rule name to not use 2012 version, since package supports 2008, 2012 and 2014 versions
+* xSQLServerNetwork
```
Comment (NarineM) [inline README.md:253]: Can you add test coverage or some E2E example for the new xSQLServerNetwork resource?

**Comment by NarineM** [comment]
Hi @Arturas-K ,
Looks like you filled out the CLA form but did not sign the document in DocuSign. Can you please sign it in DocuSign?

**Comment by Arturas-K** [comment]
Thanks @NarineM for accepting this PR. This made my day.

Since this is closed now i'll add example you asked in separate PR.

**Comment by NarineM** [comment]
@Arturas-K 
Thank you for your contributions.

**Pull request merged**